scripts/generate_changelog.js

```diff
@@ -49,36 +52,96 @@ if (!fromTag) throw Error("No process.argv[2]");
 if (!toTag) throw Error("No process.argv[3]");
 if (!outpath) throw Error("No process.argv[4]");
 
+/**
+ * @type {Record<string, {heading: string; commitsByScope: Record<string, string[]>}>}
+ */
+const sections = {
+  feat: {heading: "Features", commitsByScope: {"": []}},
+  fix: {heading: "Bug Fixes", commitsByScope: {"": []}},
+  perf: {heading: "Performance", commitsByScope: {"": []}},
+  refactor: {heading: "Refactoring", commitsByScope: {"": []}},
+  deps: {heading: "Dependencies", commitsByScope: {"": []}},
+  revert: {heading: "Reverts", commitsByScope: {"": []}},
+  build: {heading: "Build System", commitsByScope: {"": []}},
+  ci: {heading: "Continuous Integration", commitsByScope: {"": []}},
+  test: {heading: "Tests", commitsByScope: {"": []}},
+  style: {heading: "Styles", commitsByScope: {"": []}},
+  chore: {heading: "Maintenance", commitsByScope: {"": []}},
+  docs: {heading: "Documentation", commitsByScope: {"": []}},
+  _: {heading: "Miscellaneous", commitsByScope: {"": []}},
+};
+
 const isPrCommitRg = /\(#\d+\)/;
+const conventionalCommitRg = /^([a-z]+)(?:\((.*)\))?(?:(!))?: (.*)$/;
 
 const commitHashes = shell(`git log --pretty=format:"%H" ${fromTag}...${toTag}`);
 
-let commitListStr = "";
-
 for (const commitHash of commitHashes.trim().split("\n")) {
-  const subject = shell(`git log --format='%s' ${commitHash}^!`);
-  if (!isPrCommitRg.test(subject)) {
+  const rawCommit = shell(`git log --format='%s' ${commitHash}^!`);
+
+  if (!isPrCommitRg.test(rawCommit)) {
+    // Drop commits without a PR reference
     continue;
   }
 
+  const conventionalCommit = rawCommit.match(conventionalCommitRg);
+  if (!conventionalCommit) {
+    // Drop commits that do not follow conventional commit pattern
+    continue;
+  }
+
+  const [, type, scope, _breaking, subject] = conventionalCommit;
```

Comment:
Breaking changes are not highlighted at the moment, possible solution could be to have an extra section `## Breaking Changes` but based on how we used `!` so far this would be too noisy as there are too many changes marked as breaking which are not user facing.

I think if we can restrain breaking commits as highlighted by `!` to user facing changes it would provide a lot of value as it provides a quick way to just check this section to know if they are affected or not by a breaking change.

Possible solutions to highlight breaking changes

- separate Section (`## Breaking Changes`)
- highlighting in the text (`**BREAKING CHANGE:**`)
- using some kind of emoji at beginning of the line (⚠️)

Reply:
I have no strong opinion to either method. However as you already mentioned, we do have a lot of breaking changes which doesn't necessarily useful to a non-developer user/operator of Lodestar. If it does affect a user, it'll likely be highlighted already on my manual summaries.

Reply:
Let's not highlight breaking changes in the Changelog for now, until we have more clarity on it. Updated description of https://github.com/ChainSafe/lodestar/issues/5433 to discuss it there.